# Overview

The goal of this lab is to familiarize yourself with standard data wrangling tasks using Python. Being able to do so will greatly facilitate any work you do from here onward. After this lab, you should be comfortable enough in a Pythonic data environment that we can begin using a variety of packages to analyze and visualize different kinds of data more freely.

# Setting up

If you aren't already viewing these instructions from within Jupyterlab, then clone the current repository to your computer and open this document from inside Jupyter lab.

Reminder: To set up your workspace, there are several steps you should consider which we went over in previous labs...

Once you are happy with your setup, navigate to the repository you cloned and install the packages listed in the provided *requirements.txt* file:
- `pip install -r requirements.txt`. These packages include Jupyter Lab and Pandas, as well as their dependancies.

# Your second (and last) week as a data specialist at Amtrak

Your supervisor at Amtrak has handed over more "information". This time, it's a few poorly formatted tables stored as *.txt*: the most standard of file formats. They want a basic map, and since Amtrak is *finally* trying to enter the digital age with all this new federal funding, they want this map to create itself with the click of a button (in this case, by running all code cells in this notebook). Like last week, you assume there will be more data coming your way in this format, so you will need to come up with an automated process that imports it, parses it, and does any subesquent operations that are necessary in order to finally prep it for the mapping intern. In other words, you need to write a program that takes data of a certain format as input, and generates clean data as an output.

## Importing packages

At the start of any Python script (or Jupyter notebook running Python) you will usually want to import packages so that their functionality is available to the rest of your code. One of Python's great strengths is the quality of its 3rd party packages, which can provide functionality for doing anything from video editing to running a web server to... managing data! In this session, we would like to import the pandas package. Pandas allows us to manipulate data stored in pandas dataframes, and forms the bedrock of most standard data analysis done with Python today. Familiarity with other basic data management software (e.g. Excel) will serve you well here.

In [1]:
#the below command imports pandas
import pandas

#when importing a package, you can give it the alias of your choice by appending 'as' followed by your alias.
#doing so can make it more practical for you to refer to the package in your code, especially if the package were to have a really long name...
import pandas as pd

#from here on, we can refer to pandas simply as 'pd'

## Inspecting and importing data

Make sure you *know* your data. The first thing you should always do before importing data is to inspect it.
- From you command line or file explorer (Windows Explorer/Finder), open *stations.txt*...

What do you observe? Look for patterns in the data: can you spot any characteristics that you could use to parse it into a table? Ideally, you would like to convert this into a clean table where each data point is separated into its own cell under a header. If you are good in Excel, you could easily clean this up in that software. However, you want your process to be as automated as possible, so your initial script will involve some of what you did last week.

1. 📝 What would you say is the "delimiter" or "separator" for these data? Do the data have headers? Indicate your answers in the Markdown cell below.

Semicolons are used as delimiters. The data does not have headers.

- close the file *stations.txt*.

Now you're going to import this dataset. Pandas has a method called [*read_csv*](https://pandas.pydata.org/docs/user_guide/io.html#csv-text-files), which can be used to import almost any kind of table into a [pandas dataframe](https://pandas.pydata.org/docs/user_guide/dsintro.html#dataframe). Note that your data doesn't have headers, however, there is a file in your repo which indicates these...

In [2]:
#the command below is proposed for importing your table into a pandas dataframe. 
#It can take a series of optional parameters, which we call "keyword arguments"

#indicate a value for each of the two following variables,
#a string (mySep) will be passed to the 'sep' argument of the read_csv method,
#and the other, a list of strings (myHeaders), will be used by the 'names' argument. You can find these headers in headers.txt
#Refer to the read_csv documentation to understand what these arguments are doing for you...

mySep = ";"
myHeaders = ['OBJECTID_1','OBJECTID','StaType','StnType','','City','Address2','Address1','Name','Code','StationNam']
df_stations = pd.read_csv("stations.txt", sep = mySep, header = None, names = myHeaders)
#Once the data imports without error, continue!

Yay! Now you can start working with pandas.

You now have a pandas dataframe stored in a variable called "df_stations".
The print statement can be used to display the dataframe once imported, but Jupyter will render dataframes in a more appealing way if you call it without the print statement.

## Exploring pandas dataframes

You can see that not all your data is visible. It would take up a lot of space to display all your data like an excel sheet (besides, you have other software for doing that). You can use the following dataframe methods to explore your data in pandas.

In [3]:
#head will show you the top 5 rows
df_stations.head()

#you can also add a number indicating how many rows you would like to see inside the parentheses ()

OBJECTID_1  OBJECTID                               StaType StnType  \
0           1         1  Station Building (with waiting room)     BUS   
1           2         2   Curbside Bus Stop only (no shelter)     BUS   
2           3         3   Curbside Bus Stop only (no shelter)     BUS   
3           4         4  Station Building (with waiting room)   TRAIN   
4           5         6  Station Building (with waiting room)   TRAIN   

                   City               Address2                 Address1  \
0  MI48801         Alma  Transportation Center    1105 Willow Run Drive   
1  NY12211       Albany                          737 Albany Shaker Road   
2  WI54421        Colby                         1210 North Division St.   
3  MD21001     Aberdeen                          18 East Bel Air Avenue   
4  NM87102  Albuquerque                               320 1st Street SW   

                  Name Code            StationNam  
0                       AAM              Alma, MI  
1  Albany Intl Airport  ABA            Albany, NY  
2                       ABB  Abbotsford-Colby, WI  
3                       ABE          Aberdeen, MD  
4                       ABQ       Albuquerque, NM

You can also use the tail method to show the last rows of a df.

2. 📝 Use the tail method with the appropriate argument in the code cell below to show the last 15 rows of stations_df.

In [4]:
df_stations.tail()

OBJECTID_1  OBJECTID                              StaType StnType  \
1083        1084      1092  Curbside Bus Stop only (no shelter)     BUS   
1084        1085      1093  Curbside Bus Stop only (no shelter)     BUS   
1085        1086      1094  Curbside Bus Stop only (no shelter)     BUS   
1086        1087      1095  Curbside Bus Stop only (no shelter)     BUS   
1087        1088      1096           Platform only (no shelter)   TRAIN   

                                 City Address2  \
1083  CA95389  Yosemite National Park            
1084  CA95389  Yosemite National Park            
1085  CA95389  Yosemite National Park            
1086  CA95389  Yosemite National Park            
1087  AZ85364                    Yuma            

                                Address1                   Name Code  \
1083           9006 Yosemite Lodge Drive  Yosemite Valley Lodge  YOS   
1084                     Tioga Pass Road       Tuolomne Meadows  YOT   
1085                  9035 Village Drive         Visitor Center  YOV   
1086  Old Tioga Road at Highway 120 East       White Wolf Lodge  YOW   
1087                     281 Gila Street                         YUM   

                      StationNam  
1083  Yosemite National Park, CA  
1084  Yosemite National Park, CA  
1085  Yosemite National Park, CA  
1086  Yosemite National Park, CA  
1087                    Yuma, AZ

Below are some other commands you can use to explore your data...

In [5]:
#shape will return a tuple containing the number of rows followed by the number of columns.
df_stations.shape

(1088, 11)

In [6]:
#info will show you useful information related to your data
df_stations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1088 entries, 0 to 1087
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   OBJECTID_1  1088 non-null   int64 
 1   OBJECTID    1088 non-null   int64 
 2   StaType     1088 non-null   object
 3   StnType     1088 non-null   object
 4               1088 non-null   object
 5   City        1088 non-null   object
 6   Address2    1088 non-null   object
 7   Address1    1088 non-null   object
 8   Name        1088 non-null   object
 9   Code        1088 non-null   object
 10  StationNam  1088 non-null   object
dtypes: int64(2), object(9)
memory usage: 93.6+ KB


In [7]:
#describe will compute some summary statistics about your data
#note that such statistics may be meaningless based on the nature of your data
df_stations.describe() 

OBJECTID_1     OBJECTID
count  1088.000000  1088.000000
mean    544.500000   550.622243
std     314.222851   316.027068
min       1.000000     1.000000
25%     272.750000   277.750000
50%     544.500000   551.500000
75%     816.250000   824.250000
max    1088.000000  1096.000000

3. 📝 Based on the information provided by the commands above, how many stations are there in your data?

There are 1,088 stations.

4. 📝 Based on the information provided by the commands above, how many different cities are represented in your data?

1,088 cities

5. 📝 Why do you think the *describe()* method is only providing summary stats for two fields?

Because it cannot provide those statistics for object values, only interger fields can be summarized.

You can easily create new dataframes by subsetting. In other words, selecting specific columns to assign to a new dataframe. This is great for shedding columns you might not need. Subsetting a dataframe has similarities to how you would do it with a dictionary: You use the [] to select fields by their header names. See below:

In [8]:
#to isolate a single column, simply select the header like you would a key in a dictionary
#This creates a pandas 'Series', which is kind of analogous to a list. It's basically a one-dimensional dataframe...
stationTypes = df_stations['StaType']
stationTypes.head()

#pass stationTypes to the type() function. What data type is it?


0    Station Building (with waiting room)
1     Curbside Bus Stop only (no shelter)
2     Curbside Bus Stop only (no shelter)
3    Station Building (with waiting room)
4    Station Building (with waiting room)
Name: StaType, dtype: object

In [9]:
#for example, let's create a new dataframe that only contains the station code and station type
#to subset multiple columns to a new dataframe, you will need to pass a list of these headers
little_df = df_stations[['StaType','Code']]
little_df.head()

StaType Code
0  Station Building (with waiting room)  AAM
1   Curbside Bus Stop only (no shelter)  ABA
2   Curbside Bus Stop only (no shelter)  ABB
3  Station Building (with waiting room)  ABE
4  Station Building (with waiting room)  ABQ

6. 📝 In the code cell below, create a new dataframe called *temp_df* that only contains the following fields: *StaType, StnType, City, Address1*

In [10]:
temp_df=df_stations[["StaType","StnType","City","Address1"]]
temp_df.head()

StaType StnType         City  \
0  Station Building (with waiting room)     BUS         Alma   
1   Curbside Bus Stop only (no shelter)     BUS       Albany   
2   Curbside Bus Stop only (no shelter)     BUS        Colby   
3  Station Building (with waiting room)   TRAIN     Aberdeen   
4  Station Building (with waiting room)   TRAIN  Albuquerque   

                  Address1  
0    1105 Willow Run Drive  
1   737 Albany Shaker Road  
2  1210 North Division St.  
3   18 East Bel Air Avenue  
4        320 1st Street SW

Another way of shedding unneeded columns is simply to delete them in-place (within the original df). In *df_stations*, There are columns which you could do without. These would be the following: *OBJECTID_1, OBJECTID, Address2, Name*

These 4 fields are by most measures useless. You can tell because they are either zombie ID fields, or contain so few values that their use in any scenario is questionable. Search for the delete method in the [pandas documentation](https://pandas.pydata.org/docs/user_guide/index.html) and delete these fields!

(hint: you might want to check inside the page on dataframes...)

Once you find the relevant information, you will see that there are two common ways of deleting fields: 'del' and the more familiar 'pop' method used with lists.

7. 📝 Use the pop method to delete these fields.

In [11]:
#use the pop method
#then run the cell to display your dataframe below and make sure everything is as expected
df_stations.pop('OBJECTID_1')
df_stations.pop('OBJECTID')
df_stations.pop('Name')
df_stations.pop('Address2')
df_stations.head()

StaType StnType                  City  \
0  Station Building (with waiting room)     BUS  MI48801         Alma   
1   Curbside Bus Stop only (no shelter)     BUS  NY12211       Albany   
2   Curbside Bus Stop only (no shelter)     BUS  WI54421        Colby   
3  Station Building (with waiting room)   TRAIN  MD21001     Aberdeen   
4  Station Building (with waiting room)   TRAIN  NM87102  Albuquerque   

                  Address1 Code            StationNam  
0    1105 Willow Run Drive  AAM              Alma, MI  
1   737 Albany Shaker Road  ABA            Albany, NY  
2  1210 North Division St.  ABB  Abbotsford-Colby, WI  
3   18 East Bel Air Avenue  ABE          Aberdeen, MD  
4        320 1st Street SW  ABQ       Albuquerque, NM

8. 📝 Once you have run your deletions in the cell above, run the code cell again. Why does it fail? What is a KeyError? Write your answer **in bold** in the Markdown cell below.

<b>A KeyError tells us that the field inserted in the code does not exist; therefore, when we run the code again, it is telling us that the fields we are trying to delete do not exist.</b>

### Bonus question (optional):

After exploring your data in pandas for a couple minutes, you realize there is a field containing information that should probably be split! You have dealt with issues like this before...  but things might seem a little more complicated this time. Which field is it? Does it even have a header? Remind yourself of what information is contained inside this column...

How would you slice this column into two separate columns? There are many ways to do this. A familiar way might be to temporarily [turn your df into a dictionary](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_dict.html) so you can use some familiar Python methods...

In [31]:
tempDict = df_stations.to_dict()

BONUS 📝 This is an optional exercise, you are not required to fulfill it to get full points: Using tempDict, write an algorithm that will slice the state abbreviations from the zip codes. Note that the field in question does not have a header, so it can be called using an empty string. You might want to use some code from a previous lab. Some starter code is provided only for guidance...

In [ ]:
for row in tempDict['']:
    print(tempDict[''][row])
    

# Preparing your data for the mapping intern

Now that your table is all cleaned up, it's time to prepare it for mapping. There's one issue though: the data you've been working with so far doesn't seem to have any explicit location information (i.e. coordinates). There is other geographic information in your table, however, which you could potentially exploit for mapping purposes. Can you name which fields these would be?

9. 📝 List your answers in the Markdown cell below.

We can use the City, Address1 and StationNam fields

Which fields do you think you would need in order to [geocode](https://desktop.arcgis.com/en/arcmap/latest/manage-data/geocoding/what-is-geocoding.htm) these transit stations? Name all that apply in the Markdown cell below.

Address1 and StationNam is all you need to geocode. We can also use the City field, but it is redundant as the city name appears in the StationNam field as well.

Luckily, you won't need to be doing any geocoding today, since you noticed that your supervisor also left you with another file called *locs.txt*.

- Open the file in a text editor and inspect it like you did with the stations file. Do you notice a pattern here which you could use for parsing the data? 

10. 📝 Respond to the questions in the Markdown cell below:

- What would the delimiter be for these data? What file extension do we typically use to associate to this delimiter?
- Do the data have headers?
- Can you identify a field in locs.txt that you could use to merge with your stations data? Which field?

The data does not have headers and uses semicolons as the delimiter. 
The Code field can be used to merge information between locs.txt and df_stations.

11. 📝 Import *locs.txt* into Jupyterlab using the appropriate pandas method (you've done this before!) and assign the dataframe to a variable called *df_locs*.

In [14]:
df_locs=pd.read_csv('locs.txt', header= 'infer')
df_locs.head()

X          Y Code
0  -84.644837  43.391728  AAM
1  -73.809186  42.744500  ABA
2  -90.314674  44.928561  ABB
3  -76.163262  39.508455  ABE
4 -106.647985  35.082067  ABQ

## Merging dataframes

You will need to join your location data with your station information. You can do this using the pandas [merge method](https://pandas.pydata.org/docs/user_guide/merging.html#database-style-dataframe-or-named-series-joining-merging).

Before doing a merge, you should identify which column can be used to perform the merge. Merges are very similar to standard [join operations](https://dataschool.com/assets/images/how-to-teach-people-sql/sqlJoins/sqlJoins_7.png) in SQL, which you have probably seen in software like ArcGIS before (e.g. joining some data to the attribute table of a shapefile based on a common identifier).

To merge tables, you will need to identify a column that contains values which can be used to match rows from each table to eachother.

- Take note of which column you could use as a join field for merging df_stations with df_locs.

Before performing a merge, you want to make sure there are no duplicates in your data. A duplicate could mean that the same row from the table being merged gets joined twice or more. The [duplicated method](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.duplicated.html) can serve you well here.

In [15]:
#the code below will create a list of booleans, indicating True if a duplicate row is detected
#(i.e. a row is only "True" if all cells are identical to all cells contained in another row)
#and False if the row is unique.

duplicates = df_stations.duplicated()

#below is a little loop for verifying if there are any duplicates
for x in duplicates:
    if x == True:
        print(x)
    else:
        pass
    
print(x)

False


12. 📝 Are there any duplicates in df_stations? (yes or no in Markdown cell below)

No

The code below will test whether there are any duplicates specifically within the *Code* field, which could be a good field for eventually merging the data...

In [16]:
duplicates = df_stations['Code'].duplicated()

#below is a little loop for verifying if there are any duplicates
for x in duplicates:
    if x == True:
        print(x)
    else:
        pass

print(x)

False


### Using a custom function to check for duplicates

[Functions](https://www.w3schools.com/python/python_functions.asp) are a great way to organize your code. Basically, you can wrap any code inside a function, which allows you to call that code from anywhere else in your code without having to rewrite the code all over again! Think of it as a more elaborate variable, but a variable that *does* stuff, instead of just holding static information. Like variables, you need to declare a function first before using it. After that, you can refer to it anywhere, as long as you pass it the right arguments!

Functions in Python are declared using the 'def' keyword. Immediately following the name of the function - which is up to you to choose, much like any variable name - you need to add parentheses (()). Inside these parentheses is where you can pass arguments to the function (i.e. input data).

Like conditional statements or loops, function declarations are immediately followed by a colon (:), and any code nested inside a function is indented. The 'return' statement at the end of a function signals what it will output. Think of the () of the function as the point of entry, and the 'return' statement as the exit.

13. 📝 You would like to write a custom function to check for duplicates so that you can freely call it from anywhere in your code. Your function should require a **list of booleans as input**, and **return a number indicating the number of duplicates** (i.e. True values) in the list that was passed to it. Use the documentation provided above to familiarize yourself with functions. As always, you can Google for further research.

Hint: len() will provide the length of a list (and therefore a number). .append() can be used for appending True values to a list, which you can then use len() on as your return value.

In [94]:
def count_duplicates(d):
    duplicates = df_stations['City'].duplicated()
    for x in duplicates:
        if x == True:
             duplicates.add(1)
        else:
            pass
num_duplicates=count_duplicates(duplicates)
print(num_duplicates)

None


14. 📝 How many duplicates are there in the Cities field of df_stations? produce the result in the code below.

In [ ]:
I dont know...

## Merging at last!

Now that you are certain there are no duplicates, you can feel confident about merging these tables. 

15. 📝 Finally, use the documentation (links) provided above (under the **Merging dataframes** title) to perform this merge. You can also use Google if you need other resources. You will need to do your own research here. However, the task shouldn't be too complex, and you shouldn't have to use too many of the optional parameters (keyword arguments) that the merge method provides. Assign your merged data to a dataframe called *df_stations_locs*.

In [18]:
stations_locs = pd.merge(df_stations, df_locs, on="Code")

stations_locs.head()

StaType StnType                  City  \
0  Station Building (with waiting room)     BUS  MI48801         Alma   
1   Curbside Bus Stop only (no shelter)     BUS  NY12211       Albany   
2   Curbside Bus Stop only (no shelter)     BUS  WI54421        Colby   
3  Station Building (with waiting room)   TRAIN  MD21001     Aberdeen   
4  Station Building (with waiting room)   TRAIN  NM87102  Albuquerque   

                  Address1 Code            StationNam           X          Y  
0    1105 Willow Run Drive  AAM              Alma, MI  -84.644837  43.391728  
1   737 Albany Shaker Road  ABA            Albany, NY  -73.809186  42.744500  
2  1210 North Division St.  ABB  Abbotsford-Colby, WI  -90.314674  44.928561  
3   18 East Bel Air Avenue  ABE          Aberdeen, MD  -76.163262  39.508455  
4        320 1st Street SW  ABQ       Albuquerque, NM -106.647985  35.082067

---

Once you are finished, from the toolbar at the top of Jupyter lab, click Kernel > **Restart Kernel and Clear All Outputs...**. Once the notebook has completed rebooting, click Run > **Run All Cells**. Make sure that all your code runs properly, and produces your merged table at the end.

---

Fed up with being an intermediary between your supervisor's lousy data and the pesky mapping interns, you decide to take on greater challenges. You can do the mapping, and you can find quality data for doing so. Further, you have ambitions: you want to explore more interesting data that you can analyze and do research with. You pack up your bags, cross the border, and move to Montreal (on an Amtrak train)!

# Deliverables

Complete the questions in this notebook and make sure all your code runs, then push your notebook to your repo.
You will need to push the modified notebook to your repo. Make sure of the following:
- You have answered all questions and fulfilled all tasks
- Each of your code cells run properly and in order

This was the last of the introductory labs. You should so far be able to write basic python code and have a beginner's knowledge of pandas. Next week, we will dive a little deeper into pandas and start mapping our data!